File: server/datasets/java_source-code/Multiple-A2/organic7_bubble_sort.java
/*
 * Bubble Sort - organic7
 * Approach: Cocktail shaker sort - bidirectional variant of bubble sort.
 */

import java.util.Arrays;

class CocktailSort {
    public static int[] cocktailSort(int[] arr) {
        int n = arr.length;
        int start = 0;
        int end = n - 1;
        boolean swapped = true;

        while (swapped) {
            swapped = false;

            for (int i = start; i < end; i++) {
                if (arr[i] > arr[i + 1]) {
                    int temp = arr[i];
                    arr[i] = arr[i + 1];
                    arr[i + 1] = temp;
                    swapped = true;
                }
            }

            if (!swapped) {
                break;
            }

            end--;
            swapped = false;

            for (int i = end - 1; i >= start; i--) {
                if (arr[i] > arr[i + 1]) {
                    int temp = arr[i];
                    arr[i] = arr[i + 1];
                    arr[i + 1] = temp;
                    swapped = true;
                }
            }

            start++;
        }

        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {64, 34, 25, 12, 22, 11, 90, 5, 77, 1};
        System.out.println("Cocktail sorted: " + Arrays.toString(cocktailSort(arr)));
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_a2/organic7_bubble_sort.py
"""
Bubble Sort - organic7
Approach: Cocktail shaker sort - a bidirectional variant of bubble
sort that alternates passing left-to-right and right-to-left.
"""


def cocktail_sort(arr):
    n = len(arr)
    start = 0
    end = n - 1
    swapped = True

    while swapped:
        swapped = False

        for i in range(start, end):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        if not swapped:
            break

        end -= 1
        swapped = False

        for i in range(end - 1, start - 1, -1):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        start += 1

    return arr


if __name__ == "__main__":
    ARR = [64, 34, 25, 12, 22, 11, 90, 5, 77, 1]
    print("Cocktail sorted:", cocktail_sort(ARR))
